### PR TITLE
TASK: Refactor select component to be more maintainable

### PR DESCRIPTION
**Pull request by skurfuerst**
- [ ] code documentation could be improved...

### CODE REVIEW

File: packages/react-ui-components/src/ListPreviewElement/style.css [hunk 1]
```diff
@@ -0,0 +1,23 @@
+.listPreviewElement {
+    composes: reset from './../reset.css';
+    background: var(--brandColorsContrastNeutral);
+    border-top: 1px solid var(--brandColorsContrastDarker);
+    cursor: pointer;
+    text-overflow: ellipsis;
+    white-space: nowrap;
+    font-size: 14px;
```
Comment (dimaip) [inline packages/react-ui-components/src/ListPreviewElement/style.css]: font-size should be moved to reset.css eventually (but not now, it's dangerous!). Please add a TODO comment here.

File: packages/react-ui-components/src/ListPreviewElement/style.css [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    
+}
+
+.listPreviewElement--isHighlighted {
+        background: var(--brandColorsPrimaryBlue);
+}
+
+.listPreviewElement__icon {
+    composes: reset from './../reset.css';
+    padding-right: .5em;
```
Comment (dimaip) [inline packages/react-ui-components/src/ListPreviewElement/style.css]: Don't we have a styling constant for this spacing? If we don't have the right size I think we should add one more.

File: packages/react-ui-components/src/SelectBox_Header/selectBox_Header.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    render() {
+        const {
+            option,
+            showResetButton,
+            theme,
+            Icon,
+            IconButton
+        } = this.props;
+
+        // TODO: lateron, use <ListPreviewElement> here
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_Header/selectBox_Header.js]: Yes please, that would be pretty cool!

File: packages/react-ui-components/src/SelectBox_Header/style.css [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    display: inline-block;
+    width: 100%;
+}
+
+.selectBoxHeader__deleteButton {
+    composes: reset from './../reset.css';
+    position: absolute;
+    top: 0;
+    right: 41px;
+    z-index: 1;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_Header/style.css]: See how to use z-indexes now: https://github.com/neos/neos-ui/pull/1361

File: packages/react-ui-components/src/SelectBox_Header/style.css [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    text-overflow: ellipsis;
+    display: inline-block;
+    width: 100%;
+}
+
+.selectBoxHeader__deleteButton {
+    composes: reset from './../reset.css';
+    position: absolute;
+    top: 0;
+    right: 41px;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_Header/style.css]: could be `calc(var(--goldenUnit) + 1px)`, less magic numbers

File: packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/selectBox_HeaderWithSearchInput.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
+/* eslint-disable camelcase, react/jsx-pascal-case */
+import React, {PureComponent} from 'react';
+import PropTypes from 'prop-types';
+import mergeClassNames from 'classnames';
+
+const Fragment = props => props.children;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/selectBox_HeaderWithSearchInput.js]: Now that we have React 16 we could use the real fragment API?

File: packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/style.css [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        color: #fff !important;
+    }
+}
+
+
+.selectBoxHeaderWithSearchInput__loader {
+    composes: reset from './../reset.css';
+    position: absolute;
+    top: 0;
+    right: 41px;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/style.css]: same

File: packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/style.css [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    z-index: 1;
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/style.css]: same

File: packages/react-ui-components/src/SelectBox_ListPreview/selectBox_ListPreview.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+/* eslint-disable camelcase, react/jsx-pascal-case */
+import React, {PureComponent} from 'react';
+import PropTypes from 'prop-types';
+
+const Fragment = props => props.children;
+
+/**
+ * **SelectBox_ListPreview is an internal implementation detail of SelectBox**, meant to improve code quality.
+ *
+ * It is used inside SelectBox as the dropdown part.
```
Comment (dimaip) [inline packages/react-ui-components/src/SelectBox_ListPreview/selectBox_ListPreview.js]: But could also be used in the header, for selected element, right?

Comment (skurfuerst) [reply packages/react-ui-components/src/SelectBox_ListPreview/selectBox_ListPreview.js]: Nope, there, the ListPreviewElement should be used directly I think.

File: packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/selectBox_HeaderWithSearchInput.js (hunk 6 above)
Comment (skurfuerst) [reply packages/react-ui-components/src/SelectBox_HeaderWithSearchInput/selectBox_HeaderWithSearchInput.js]: We need React 16.2.0 at least: https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html

**Comment by kitsunet** [comment]
I removed the WIP status. But this needs extensive testing and review.

**Comment by kitsunet** [comment]
Mmm, I am not sure I can fix the remaining tests. This is too obscure for me without background.

**Comment by kitsunet** [comment]
dmitri fixed some of the test errors. The reminder seems to evolve around "loading" messages. We should decide if we want to drop them and adjust the tests or if we add it back in.

File: packages/react-ui-components/src/SelectBox/selectBox.js [hunk 10]
```diff
@@ -185,6 +185,7 @@ export default class SelectBox extends PureComponent {
                         <SelectBox_ListPreview
                             {...this.props}
 
+                            theme={theme}
```
Comment (skurfuerst) [inline packages/react-ui-components/src/SelectBox/selectBox.js]: Is this actually needed / correct? I think we never should need this; but the corresponding styles should be in SelectBox_ListPreview/style.css then

**Pull request merged**